File: javascript/270.js
const randint = (val) => {
  return Math.floor(Math.random() * val);
};
class TreeNode {
  constructor(val) {
    this.val = val;
    this.left = null;
    this.right = null;
  }
}

const random_BST_rec = () => {
  let len = randint(30);
  console.log(len);
  let root = new TreeNode(randint(50) + 30);

  const walk = (node) => {
    node.left = new TreeNode(node.val - randint(10));

    node.right = new TreeNode(node.val + randint(10));
    if (Math.random() > 0.5) {
      walk(node.left);
    }
    if (Math.random() > 0.5) {
      walk(node.right);
    }
  };
  walk(root);
  return root;
};
const inorder_traverse = (root) => {
  let arr = [];

  const walk = (node) => {
    if (node === null) {
      return;
    }
    walk(node.left);
    arr.push(node.val);
    walk(node.right);
  };
  walk(root);
  // console.log(arr);
  return arr;
};
const closestValue = (root, target) => {
  let min = Number.MAX_VALUE;
  let res = -1;
  const walk = (node) => {
    if (node === null) {
      return;
    }
    if (Math.abs(node.val - target) < min) {
      min = Math.abs(node.val - target);
      res = node.val;
    }
    walk(node.left);
    walk(node.right);
  };
  walk(root);
  return res;
};
//--------------------main----------------------------------------------------
// let root = random_BST();
// console.log(root);
let root = random_BST_rec();
// console.log(root);
console.log(inorder_traverse(root));
let target = Math.random() * randint(100);
console.log(closestValue(root, target));

File: python/270.py
# Given the root of a binary search tree and a target value, return the value in the BST that is closest to the target.

 

# Example 1:


# Input: root = [4,2,5,1,3], target = 3.714286
# Output: 4
# Example 2:

# Input: root = [1], target = 4.428571
# Output: 1
 

# Constraints:

# The number of nodes in the tree is in the range [1, 104].
# 0 <= Node.val <= 109
# -109 <= target <= 109


# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution(object):
    def scan(self,node):
        if node==None:
            return
        temp=abs(node.val-self.t)
        if temp<self.min:
            self.min=temp
            self.res=node.val
        self.scan(node.left)
        self.scan(node.right)
    def closestValue(self, root, target):
        """
        :type root: TreeNode
        :type target: float
        :rtype: int
        """
        self.t=target
        self.min=9999999
        self.res=0
        self.scan(root)
        return self.res
